### DISCUSSION TITLE
Should a descriptor (with `__set__`) marked as "final" allow assignments from instances?

### DISCUSSION THREAD

**Discussion by bzoracler** [python/typing#1756] [Q&A]
My intuition is "yes", based on the behaviour of `@property` ([mypy Playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=0c848ed23f2b4b48cf45f38633476d77), [pyre Playground](https://pyre-check.org/play?input=import%20typing%20as%20t%0A%0Aclass%20A%3A%0A%20%20%20%20%40property%0A%20%20%20%20%40t.final%0A%20%20%20%20def%20prop(self)%20-%3E%20int%3A%0A%20%20%20%20%20%20%20%20return%200%0A%20%20%20%20%0A%20%20%20%20%40prop.setter%0A%20%20%20%20%40t.final%0A%20%20%20%20def%20prop(self%2C%20val%3A%20int%2C%20%2F)%20-%3E%20None%3A%20...%0A%0A%0Aa%3A%20A%20%3D%20A()%0Aa.prop%20%3D%208%0Aa.prop%20%3D%20%22%22%0A), [Pyright playground](https://pyright-play.net/?strict=true&code=JYWwDg9gTgLgBDAnmYA7A5nAhgZwQKHwGMAbXPAQQC58464ABMKCMAU1kVvoZgDoAZmiwludACZsBcZqwAUONiQEBKOAFoAfHDQwa9A3ChsYAVyio4ABjFxbTFmD6KYMDvf5DUI25OmywBSUBABo4ADcRKh1UGDCAejUtOAA5CFQ2aL5swnwsaIo4AF44CjkVPL4A4rgADkrqkoAiJvwgA)):

```python
import typing as t

class A:
    @property
    @t.final
    def prop(self) -> int:
        return 0
    
    @prop.setter
    @t.final
    def prop(self, val: int, /) -> None: ...


a: A = A()
a.prop = 8   # OK
a.prop = ""  # Type error
```

However, the type-checkers tested above don't seem to support custom final descriptors participating in instance assignments ([mypy Playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=c193368684af234df9a4055dbd054ff7), [pyre Playground](https://pyre-check.org/play?input=import%20typing%20as%20t%0A%0Aclass%20D%3A%0A%20%20%20%20def%20__set__(self%2C%20obj%3A%20t.Any%2C%20val%3A%20int%2C%20%2F)%20-%3E%20None%3A%20...%0A%0Aclass%20A%3A%0A%20%20%20%20desc%20%3D%20D()%0A%20%20%20%20desc_final%3A%20t.Final%20%3D%20D()%0A%0Aa%3A%20A%20%3D%20A()%0Aa.desc%20%3D%208%0Aa.desc%20%3D%20%22%22%0Aa.desc_final%20%3D%208%0Aa.desc_final%20%3D%20%22%22), [Pyright Playground](https://pyright-play.net/?strict=true&code=JYWwDg9gTgLgBDAnmYA7A5nAhgZwQKHwGMAbXPAEQC58464ATAUwDM4B9dnJmTgCm4kWAGjgQARgCsqCAHQBBVIlEA3LCRloYogPQBKOAFoAfHAByEVExmzbhUuTjya9RkxxE4AXjgU%2Be2npmD3YWNHUZGFkAMXCSb19-QiwZeQT5JKxZYM8fAA58LJyEgCISwuz3IlC4hIKiqprUdVKSoA)):

```python
import typing as t

class D:
    def __set__(self, obj: t.Any, val: int, /) -> None: ...

class A:
    desc = D()
    desc_final: t.Final = D()

a: A = A()
a.desc = 8         # OK
a.desc = ""        # Expected type error
a.desc_final = 8   # Type-checkers complain here about assignment
a.desc_final = ""  # Type-checkers complain here about assignment
```

**Comment by Daverball** [comment]
This is just completely unspecified behavior. `typing.final` is not supposed to work with descriptors, only with methods and classes, and indeed mypy complains about it. Extending `typing.final` to work with `property` and other decorators that return arbitrary types seems like it could make sense, but there are some nuances about it that make it non-trivial. Namely type checkers disallow assignments to methods, but not necessarily for some of the other types that could be returned by the descriptor.

The behavior with `typing.Final` makes complete sense, since it always applies to the parent scope, rather than the enclosed type. Changing this behavior based on the RHS seems like a bad idea to me, because what do you with `Any`?

There will always be some ambiguity about what final exactly means here and what it applies to, but I think the current interpretation is the least likely to cause consistency issues, so I think the only reasonable way to extend `typing.final` to work with decorators that return arbitrary objects is to make it equivalent to `typing.Final` i.e. the following two things are equivalent:

```python
class Foo:
    x: Final = desc(lamda: ...)
    @final
    @desc
    def y(self) -> int: ...
```

Which is not what you want. But I think that if we want to be able to create descriptors that are settable but can't be overridden in subclasses we need a new, separate type qualifier.

**Reply by bzoracler** [reply]
> Changing this behavior based on the RHS seems like a bad idea to me, because what do you with `Any`?

Thanks for bringing up this up, if you mean the potential of false positives in this situation:

```python
class Foo:
    x: Final[Any] = desc_with__set__()

foo: Foo = Foo()
foo.x = 1  # Should this be a type-checker error?
```

`Final` here is not the main issue - rather, it is the appearance of `x` marked explicitly as a class-scoped variable, so `x: ClassVar[Any]` would suffer the same potential ambiguity. Pyright handles this by indeed marking the assignment as an error (see [Pyright Playground example](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAMIA2AhgM60Bq1IAUCwMY31QAmAprXYB9AO5IYACyFDafGNIBcLKCt59gUabPlCAFLMrAyYAEYArBYWJkAbtUqWipKAHoAlFAC0APigA5MBQ%2BSwA6MLZOOlooADEwMCVVKAAPAEZLKiimECgAXjVBUXEpGTlpXTdlVWSAJgyuRmYAbScAXTyC4TFJLTK9SqTkgGZ6rObUfAAfTqKe0p12-P5C7pLtcsqWYHjLOLAOvYqt%2BJC0jtTjsFOa88vTodugA), also https://github.com/microsoft/pyright/issues/4837).

I agree with the way Pyright handles this. My interpretation is that `Any` is only concerned with type compatibility, not scope compatibility. Without knowledge of what `Foo.x` is, `foo.x` should be interpreted as an instance-level assignment, and I think it is out-of-scope (!) for a class-scoped `Any` to consider assignment compatibility at the instance level. In this example, then, a class-scoped `Any` should not widen the possibility of the instance-level `foo.x = 1` to be interpreted as `type(foo).x.__set__(foo, 1)`, and `foo.x = 1` should be continued (as currently) to be treated as an error.

If `Any` is the only ambiguous factor here, I don't think this is necessary:

> ... I think that if we want to be able to create descriptors that are settable but can't be overridden in subclasses we need a new, separate type qualifier.

**Reply by Daverball** [reply]
Making `Final`/`ClassVar` only apply to the class level depending on whether or not that specific attribute contains a descriptor or not contradicts the documentation and its behavior is currently not explicitly specified in the typing spec, so I think you will need to open a discussion at https://discuss.python.org/ and subsequently create pull request here that amends the typing spec accordingly, if you want to change what type checkers do here.

I'm worried about adding another layer of confusion on top of mutability vs. assignability with descriptors having their own rules. I'd rather make the type system more expressive here, so each case can be covered unambiguously, even if there is a case to be made for special-casing descriptors.

See the README here for a concrete list of steps to follow: https://github.com/python/typing-council
